File: archive/pulumi-py/Pr1217/lib/components/iam_roles.py
import json
import pulumi
import pulumi_aws as aws
from pulumi import Output


class IAMRoles(pulumi.ComponentResource):
  """
  Component for creating IAM roles and policies following least privilege principle.
  Also creates a TestHarnessRole for integration testing.
  """

  def __init__(self, name: str, bucket_arn: Output[str], opts: pulumi.ResourceOptions = None):
    super().__init__('custom:aws:IAMRoles', name, {}, opts)

    lambda_assume_role_policy = json.dumps({
      "Version": "2012-10-17",
      "Statement": [
        {
          "Action": "sts:AssumeRole",
          "Effect": "Allow",
          "Principal": {
            "Service": "lambda.amazonaws.com"
          }
        }
      ]
    })

    self.lambda_role = aws.iam.Role(
      f"{name}-lambda-role",
      assume_role_policy=lambda_assume_role_policy,
      description="IAM role for Lambda function execution",
      tags={
        "Name": f"{name}-lambda-role",
        "Component": "Lambda"
      },
      opts=pulumi.ResourceOptions(parent=self)
    )

    # Attach basic Lambda execution policy
    self.lambda_basic_execution_attachment = aws.iam.RolePolicyAttachment(
      f"{name}-lambda-basic-execution",
      role=self.lambda_role.name,
      policy_arn="arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
      opts=pulumi.ResourceOptions(parent=self)
    )

    # Lambda S3 Access Policy
    self.s3_access_policy = aws.iam.Policy(
      f"{name}-s3-access-policy",
      description="Policy for Lambda to access S3 bucket",
      policy=bucket_arn.apply(lambda arn: json.dumps({
        "Version": "2012-10-17",
        "Statement": [
          {
            "Effect": "Allow",
            "Action": [
              "s3:GetObject",
              "s3:GetObjectVersion",
              "s3:GetBucketNotificationConfiguration",
              "s3:HeadObject"
            ],
            "Resource": [f"{arn}/*", arn]
          }
        ]
      })),
      opts=pulumi.ResourceOptions(parent=self)
    )

    self.s3_policy_attachment = aws.iam.RolePolicyAttachment(
      f"{name}-s3-policy-attachment",
      role=self.lambda_role.name,
      policy_arn=self.s3_access_policy.arn,
      opts=pulumi.ResourceOptions(parent=self)
    )

    self.register_outputs({
      'lambda_role_arn': self.lambda_role.arn,
      'lambda_role_name': self.lambda_role.name
    })

File: lib/src/main/java/app/components/IAMRoles.java
package app.components;

import com.pulumi.aws.Provider;
import com.pulumi.aws.iam.Role;
import com.pulumi.aws.iam.RoleArgs;
import com.pulumi.aws.iam.RolePolicyAttachment;
import com.pulumi.aws.iam.RolePolicyAttachmentArgs;
import com.pulumi.core.Either;
import com.pulumi.core.Output;
import com.pulumi.resources.ComponentResource;
import com.pulumi.resources.ComponentResourceOptions;
import com.pulumi.resources.CustomResourceOptions;

public class IAMRoles extends ComponentResource {
    private final Output<String> administrationRoleArn;
    private final Output<String> executionRoleName;

    public IAMRoles(String name, Provider provider) {
        super("custom:aws:IAMRoles", name, ComponentResourceOptions.builder().provider(provider)
                .build());

        // StackSet Administration Role
        var administrationRole = new Role("stackset-administration-role", RoleArgs.builder()
                .name("AWSCloudFormationStackSetAdministrationRole")
                .assumeRolePolicy("""
                        {
                            "Version": "2012-10-17",
                            "Statement": [
                                {
                                    "Effect": "Allow",
                                    "Principal": {
                                        "Service": "cloudformation.amazonaws.com"
                                    },
                                    "Action": "sts:AssumeRole"
                                }
                            ]
                        }
                        """)
                .build(), CustomResourceOptions.builder().parent(this).provider(provider).build());

        // Attach policy to administration role
        new RolePolicyAttachment("stackset-admin-policy-attachment",
                RolePolicyAttachmentArgs.builder()
                        .role(administrationRole.name())
                        .policyArn("arn:aws:iam::aws:policy/service-role/AWSCloudFormationStackSetAdministrationRole")
                        .build(), CustomResourceOptions.builder()
                .parent(this)
                .provider(provider)
                .build());

        // StackSet Execution Role (to be created in target accounts)
        var executionRole = new Role("stackset-execution-role", RoleArgs.builder()
                .name("AWSCloudFormationStackSetExecutionRole")
                .assumeRolePolicy(administrationRole.arn().applyValue(adminArn -> Either.ofLeft(String.format("""
                        {
                            "Version": "2012-10-17",
                            "Statement": [
                                {
                                    "Effect": "Allow",
                                    "Principal": {
                                        "AWS": "%s"
                                    },
                                    "Action": "sts:AssumeRole"
                                }
                            ]
                        }
                        """, adminArn))))
                .build(), CustomResourceOptions.builder().parent(this).provider(provider)
                .build());

        // Attach comprehensive policy to execution role
        new RolePolicyAttachment("stackset-exec-policy-attachment",
                RolePolicyAttachmentArgs.builder()
                        .role(executionRole.name())
                        .policyArn("arn:aws:iam::aws:policy/PowerUserAccess")
                        .build(), CustomResourceOptions.builder()
                .parent(this)
                .provider(provider)
                .build());

        this.administrationRoleArn = administrationRole.arn();
        this.executionRoleName = executionRole.name();
    }

    public Output<String> getAdministrationRoleArn() {
        return administrationRoleArn;
    }

    public Output<String> getExecutionRoleName() {
        return executionRoleName;
    }
}
